build_functions/CMakeLists.txt

```diff
@@ -25,36 +33,82 @@ function(build_lib whatIsBuilding solution_folder)
-    set(ARG_PREFIX "none")
-    foreach(f ${ARGN})
-        set(skip_to_next FALSE)
-        if(${f} STREQUAL "ADDITIONAL_LIBS")
-            SET(PARSING_ADDITIONAL_LIBS ON)
-            SET(PARSING_VALGRIND_SUPPRESSIONS_FILE OFF)
-            set(ARG_PREFIX "none")
-            #also unset all the other states
-            set(skip_to_next TRUE)
-        elseif(${f} STREQUAL "VALGRIND_SUPPRESSIONS_FILE")
-            SET(PARSING_ADDITIONAL_LIBS OFF)
-            SET(PARSING_VALGRIND_SUPPRESSIONS_FILE ON)
-            set(skip_to_next TRUE)
-        endif()
+    set(options) #there are no options
+    set(oneValueArgs) #no single value arg
+    set(multiValueArgs ADDITIONAL_LIBS MOCK_PRECOMPILE_HEADERS NO_MOCK_PRECOMPILE_HEADERS) # only ADDITIONAL_LIBS is a multi value arg
 
-        if(NOT skip_to_next)
-            if(PARSING_ADDITIONAL_LIBS)
-                if((${f} STREQUAL "debug") OR (${f} STREQUAL "optimized") OR (${f} STREQUAL "general"))
-                    SET(ARG_PREFIX ${f})
-                else()
-                    target_link_libraries_with_arg_prefix(${ARG_PREFIX} ${whatIsBuilding}_lib_${CMAKE_PROJECT_NAME} ${f})
-                    set(ARG_PREFIX "none")
-                endif()
+    cmake_parse_arguments("arg" "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN}) #"arg" is the prefix added to variables detected by cmake_parse_arguments
+
+    if(arg_ADDITIONAL_LIBS)
+        set(ARG_PREFIX "none") # "none" results in a link without prefix in function target_link_libraries_with_arg_prefix. Also see debug, optimized, or general keywords in https://cmake.org/cmake/help/latest/command/target_link_libraries.html
+        foreach(f ${arg_ADDITIONAL_LIBS})
+            if((${f} STREQUAL "debug") OR (${f} STREQUAL "optimized") OR (${f} STREQUAL "general"))
+                SET(ARG_PREFIX ${f})
+            else()
+                target_link_libraries_with_arg_prefix(${ARG_PREFIX} ${whatIsBuilding}_lib_${CMAKE_PROJECT_NAME} ${f})
+                set(ARG_PREFIX "none")
             endif()
-        endif()
+        endforeach()
+    else()
+        #arg_ADDITIONAL_LIBS is not set therefore no need to do anything about libraries
+    endif()
+
+    if(arg_MOCK_PRECOMPILE_HEADERS OR arg_NO_MOCK_PRECOMPILE_HEADERS)
+
+        string(TOUPPER ${whatIsBuilding} LIBRARY_NAME_UPPER)
+
+        set(MOCKS_HEADER_FILE ${CMAKE_CURRENT_BINARY_DIR}/${whatIsBuilding}_mocks.h)
+        set(MOCKS_SOURCE_FILE ${CMAKE_CURRENT_BINARY_DIR}/${whatIsBuilding}_mocks.c)
+
+        set(MOCKS_HEADER_CONTENT "// Copyright (c) Microsoft. All rights reserved.\n")
+        set(MOCKS_HEADER_CONTENT "${MOCKS_HEADER_CONTENT}// Licensed under the MIT license. See LICENSE file in the project root for full license information.\n")
+        set(MOCKS_HEADER_CONTENT "${MOCKS_HEADER_CONTENT}// THIS IS A GENERATED FILE, DO NOT EDIT. It was generated using function build_lib from CMakeLists.txt in testrunnerswitcher.\n")
```

Comment:
>nnerswitche [](http://example.com/codeflow?start=170&length=11)

Consider putting the link to the CMakeList.Txt :-) Easy to find.

Reply:
added http link